### DISCUSSION TITLE
Display of a booking form field based on the value of a person-type custom field

### DISCUSSION THREAD

**Discussion by f14m07** [liedekef/events-made-easy#541] [Q&A] [answered]
Hi,
The custom field{23} is field type 'Text' and field type purpose 'People field'.
I would like to display a booking form field based on the value of a person-type custom field{23} 
As the code shown below the third condition is not taken into accounts. What is wrong? Is there an other solution ?
```
[eme_if tag='#ESC_PRICE' value='25']
                 [eme_if2 tag='#_IS_USER_MEMBER_OF{Cartes}' value='1'] 
                    [eme_if3 tag='#ESC_FIELD{23}' notincsv='0,1']        
                       <tr><th>Voulez-vous utiliser votre carte de réduction ?<br></th>
                       <td>#_FIELD{03}coût= 2 coupons, [eme_mymemberinfo membership_id=1 template_id=85]
                    </td></tr>
                    [/eme_if3]
                 [/eme_if2]
            [/eme_if]
```

**Comment by f14m07** [comment]
I got round my difficulty by storing the people concerned in a dynamic group and using #_IS_USER_IN_GROUP{xxx}.
I hadn't realized that this new dynamic group option was available.

**Reply by liedekef** [reply]
Well, doing conditions on custom people fields in a form is not easy to handle. Currently it won't replace the logged-in user info as you'd expect, because what if you allow the value to be changed (so in fact it doesn't correspond to the logged-in user)? So I would need to check a lot more info (the event restrictions) before replacing #_FIELD with its value, and that is not coded up right now.

**Comment by f14m07** [comment]
I have an other issue. When I create the dynamic group I choose the value '1' but I have in this group also those who have the value 11,12,13 etc

**Reply by liedekef** [reply]
That's because the search for custom field values is an open text search ("anything that contains"), not an exact search.

**Reply by f14m07** [reply]
How can I specify just one character?

**Reply by liedekef** [reply]
That requires quite some code changes ...

**Reply by f14m07** [reply]
If I want to exclude only those with the value 1, can I create dynamic groups with 11,12,13,14,15,16,17,18,19? Can I reverse my test in the booking form? This should be possible if there is no limit to the number of groups I can enclose between {}.

**Reply by liedekef** [reply]
Your question made me look at the "exact search" match addition and while I was at it, I also discovered a bug when directly adding dynamic groups (via the "groups" submenu).
The dev-version has all the fixes, feel free to try it.

**Reply by f14m07** [reply]
I upload the dev version on my local site. In the admin screen, when I filter on the value '0' on my custom field. The values '0' and '20' or '10' are selected. Not I had expected.

**Reply by liedekef** [reply]
Did you create a filter with "exact match" checked? If not, the search behaviour remains as it was before.

**Reply by liedekef** [reply]
Correction: when editing a dynamic group, it will do as said, but when searching as is in the people overview, it is not taken into account in the ajax call. I'll fix that

**Reply by liedekef** [reply]
The current dev version contains js fixes now (make sure to do a ctrl-reload for the cache to be refreshed)

**Reply by f14m07** [reply]
Ok, no cache on my local. My custom field to filter is a member custom field. On the member screen it works fine. But the dynamic group has a bug because the option "Exact" is not respected

**Reply by liedekef** [reply, marked as answer]
indeed, more things (it is still in test :-) )
The latest dev should fix this (also group edit lost the dynamic property, should be fixed too)

**Comment by f14m07** [comment]
Hi, I've downloaded the development version and everything looks fine from my end.

**Comment by f14m07** [comment]
I've also downloaded the dev version to the prod site. The behavior is not quite the same. The dynamic group is correct, but you have to remember to select "Exact", which was not memorized when the group was created from the "Member" screen. The selection on the 'Member' screen doesn't work properly; the 'Exact' tick box doesn't seem to be taken into account.

**Comment by f14m07** [comment]
In fact, there's a problem with the dynamic group too. Only the number of people is correct, but in fact the group contains extra people who shouldn't be there.
I don't understand why I'm not behaving correctly, unless I've installed the wrong dev version. I'll try it again. The cache has been empty.

**Comment by f14m07** [comment]
I have found a difference between the local and prod site. On the first one the type of the custom member field is 'texte' and it was 'Number(HTML5)' on the prod. Could this be why things work differently?
However, in the database I saw that the 'answer' field in the eme_answers table was of type 'text'.
I'm not sure what to do

**Reply by liedekef** [reply]
in the answer table all fiels are of type text, but unless you describe your setup and issue in more detail, I can't help of course :-)
I'll test further at home once I find the time (2 days)

**Reply by f14m07** [reply]
On the member admin screen, you can see the column with 0 and 20. Exact is not taken.
<img width="1115" alt="image" src="https://github.com/liedekef/events-made-easy/assets/113473805/64b911f6-ed28-42f2-b9fd-0232b433f844">

**Comment by f14m07** [comment]
Hi, 
My problem is that the local site and the production site have the same version of EME but don't behave exactly the same. The difference between my local and the prod is that the "Exact" checkbox is not taken into account on the site in prod.
I thought it might be due to a PHP module not being loaded on the server? Here are the options I can check or uncheck.Do you see a pb?
<img width="806" alt="image" src="https://github.com/liedekef/events-made-easy/assets/113473805/a5381a86-0e2c-4133-9ca0-6842d504407b">

Could the problem be with another plugin? Do you accept to log as admin?

**Reply by liedekef** [reply]
The php modules are not related. Are the changed javascript files copied to prod too? Feel free to send me an admin login.

**Reply by f14m07** [reply]
Yes the prod is with the dev version js included
You have already an access on https://vogue-et-reve.fr with your adress f****y@e-********.be

**Reply by liedekef** [reply]
Something is caching the javascript on your site:
https://vogue-et-reve.fr/wp-content/plugins/events-made-easy/js/eme_admin_members.js?ver=2.4.42 gives the wrong code while
https://vogue-et-reve.fr/wp-content/plugins/events-made-easy/js/eme_admin_members.js is ok.

**Reply by f14m07** [reply]
So what should I do?
I don't have the right version (Dev) installed?
I need to correct my cache?

**Reply by f14m07** [reply]
Do you say I have not the right version of eme_admin_member.js?
Or do you say I have the right one but something (plugin?) is preventing the java script from working?

**Reply by f14m07** [reply]
So I've downgraded my cache plugin a bit more, and it's working now.
Thanks for the development work. I've done a lot of searching because some things aren't obvious to me

**Reply by liedekef** [reply]
I never recommend a caching plugin ...

**Reply by liedekef** [reply]
A caching plugin tries to do what exactly never should be done: cache dynamically generated pages. This is not the first time a caching plugin caches wordpress nonces that then result in forms not being able to submit.
If you need caching, do this via your webserver (to cache javascript files, images, css files). If you want some kind of performance gain in WP itself, use a database caching plugin that is intelligent enough to invalidate its cache if something gets updated. Even WP itself can do object caching via either in mem caching, memcached or redis (and even allows a developer to hook into that, even I do it).
